### ISSUE TITLE
What if you already have the original bandwidthd running?

### ISSUE THREAD

**Issue by jult** [perkinslr/bandwidthd-ipv6#1]
Do I uninstall first, or what do you advice? Can it read existing cdf logs?

**Comment by jult** [comment]
Never mind, it's a complete disaster even compiling it on debian 12. Pity this never got updated.

**Comment by tlhonmey** [comment]
Yeah, the original and all the forks explode messily when you try to compile them on modern systems.  I'm not sure if this one or the OpenWRT fork are closer to being functional.  You get to choose between automake macros that are obsolete, or the linker pitching a fit over some of the inline functions...

**Comment by perkinslr** [comment]
I haven't had need of this in over 10 years, and I haven't looked at it in nearly as long.  It's only about 3500 lines of C, with a bit of bash and php for rendering and maintenance.  It's likely to be easier to rewrite it than port this forward, but shouldn't be too much work to do.  If I get some free time in the next couple weeks, I'll take a shot at rewriting it in modern c++ and pybind11, with backwards compatibility.

**Comment by perkinslr** [comment]
Hm, making it work on openwrt will be a bit annoying, as that pretty well precludes using python to talk to the database or handle configuration parsing.  That level of stuff might be left as an exercise for the user.

**Comment by jult** [comment]
> I haven't had need of this in over 10 years, and I haven't looked at it in nearly as long.

What have you used instead? Or do you mean you had no need to monitor bandwidth, regardless?
Either way, would be great if this wonderful little gem of a tool could function again, on a debain 12 router. I would donate for it!

**Comment by perkinslr** [comment]
I left the job where I needed to monitor bandwidth long about 2014; this has been left to bitrot since.  I'm mostly through rewriting it, the core is super simple, it's all the layers around it that make it easy to use that are fiddly. 

 I'm going to see about setting up Actions here that should be able to provide deb packages and maybe see if I can static compile it for MIPS and some of the other low-end router CPUs. That'll probably take a _little_ refactoring, since I don't think uclibc++ has std::format available.

 That said, if there are any particular features you wish worked a little differently, now would be the time to address them. For example, I'm undecided on if and how to embed graph drawing. The original used a png drawing library, but that's heavyweight if the goal is to run it on OpenWRT. If you're running it on a full system (Debian 12 or similar), then there's no real need to have it draw its *own* graphs, as outsourcing the graph creation to graphviz / PIL / matplotlib is much easier to customize, and unify with graphs from the database.  I could have it output raw bitmaps, but I'm not sure if it's worth the time to write that.  

Anyway, I'll probably update the readme here to point to a new repo under my company's account, and then archive this one.    If you were serious about donating, even simply to buy me a coffee, I'll have that included in the new repo.

Edit:  Yup, uClibc++ is missing lots of modern niceties. Wouldn't be *that* hard to compile with it, but I think I'll not bother till someone *needs* it to fit on a tiny system.

**Comment by perkinslr** [comment]
This has been sitting on my machine for probably the last 4 months, in a usable but *not-quite-ready* state. I finally had a bit of time to spare, and got it into shippable shape.  You can find it [here](https://github.com/lp-programming/bandwidthd-ng). 

@jult You mention recovering from cdfs.  That's not a feature I've ever *used*, since I always use it with a database.  I can add it in short order if you have a use case for it that sqlite doesn't just handle, but I'd rather not do the work of adding it if there isn't actually a use for it. 

Anyway, you want it running on debian 12 (oldstable). That should be pretty easy to do.  I can furnish a package for it, or theoretically the make_deb.py script should be able to do it (you'd need libc++-19-dev and a relatively recent version of clang on the target machine).  In theory, the portable build artifacts also should work anywhere with a modern glibc and libpq installed. I can also make a postgresless portable version on request.

**Comment by perkinslr** [comment]
@jult I can't mention you in the other repo, so I'm linking it here:

https://github.com/lp-programming/bandwidthd-ng/issues/1

I can walk you through building it for debian 12, or you're welcome to give the built package a spin to ensure it works. You'll need to install postgres, libpcap, and llvm-19/libcxx-19 and possibly a couple others.
